README.md
# DEPRECATED
Use [dynamic_color](https://pub.dev/packages/dynamic_color) by the Material team.

## Monet Colors

A Flutter plugin that exposes [Monet (Material You, Material 3)](https://m3.material.io/styles/color/dynamic-color/user-generated-color) [system colors](https://developer.android.com/reference/android/R.color#system_accent1_0) on Android 12. Returns `null` on unsupported platforms and leaves handling that to you. For a backwards compatible implementation, see [monet](https://pub.dev/packages/monet).

![](https://i.imgur.com/1sQEPAb.png)

## Install & Use

See: [Installing](https://pub.dev/packages/monet_colors/install).

```dart
final MonetColors? colors = await Monet.get();

final primary200 = colors.primary[200];
final secondary400 = colors.secondary.shade400;
final tertiary = colors.tertiary;
final tertiary300 = tertiary[300];
// ...
```

Check out the example project for a demo of all the colors.


CHANGELOG.md
## 0.9.1

Deprecation :)

## 0.9.0

Initial release

## 0.0.1


test/monet_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('monet');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    // expect(await Monet.platformVersion, '42');
  });
}


lib/src/colors.dart
import 'package:flutter/material.dart';

class MonetColor extends ColorSwatch<int> {
  final Map<int, Color> colors;

  MonetColor._(this.colors)
      : assert(colors.length == 13),
        super(colors[500]!.value, colors);

  /// White.
  Color get shade0 => this[0]!;

  /// The lightest shade.
  Color get shade10 => this[10]!;

  /// The second lightest shade.
  Color get shade50 => this[50]!;

  /// The third lightest shade.
  Color get shade100 => this[100]!;

  /// The fouth lightest shade.
  Color get shade200 => this[200]!;

  /// The fifth lightest shade.
  Color get shade300 => this[300]!;

  /// The sixth lightest shade.
  Color get shade400 => this[400]!;

  /// The default shade.
  Color get shade500 => this[500]!;

  /// The fourth darkest shade.
  Color get shade600 => this[600]!;

  /// The third darkest shade.
  Color get shade700 => this[700]!;

  /// The second darkest shade.
  Color get shade800 => this[800]!;

  /// The darkest shade.
  Color get shade900 => this[900]!;

  /// Black.
  Color get shade1000 => this[1000]!;
}

class MonetColors {
  /// The primary accent color.
  final MonetColor primary;

  /// The secondary accent color.
  final MonetColor secondary;

  /// The tertiary accent color.
  final MonetColor tertiary;

  /// The neutral color.
  final MonetColor neutral;

  /// The neutral variant color.
  final MonetColor neutralVariant;

  MonetColors({
    required this.primary,
    required this.secondary,
    required this.tertiary,
    required this.neutral,
    required this.neutralVariant,
  });

  /// Creates an instance from a map of raw color names and int colors taken
  /// from the system.
  factory MonetColors.fromMap(Map<String, int> colors) {
    final accent1 = <int, Color>{};
    final accent2 = <int, Color>{};
    final accent3 = <int, Color>{};
    final neutral1 = <int, Color>{};
    final neutral2 = <int, Color>{};

    colors.forEach((name, value) {
      /// Example color name: accent1_500
      final type = name.split('_')[0];
      final shade = int.parse(name.split('_')[1]);

      switch (type) {
        case 'accent1':
          accent1[shade] = Color(value);
          break;
        case 'accent2':
          accent2[shade] = Color(value);
          break;
        case 'accent3':
          accent3[shade] = Color(value);
          break;
        case 'neutral1':
          neutral1[shade] = Color(value);
          break;
        case 'neutral2':
          neutral2[shade] = Color(value);
          break;
      }
    });

    return MonetColors(
      primary: MonetColor._(accent1),
      secondary: MonetColor._(accent2),
      tertiary: MonetColor._(accent3),
      neutral: MonetColor._(neutral1),
      neutralVariant: MonetColor._(neutral2),
    );
  }
}


android/src/main/kotlin/dev/ihsan/monet/MonetPlugin.kt
package dev.ihsan.monet

import android.content.Context
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class MonetPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel

  private lateinit var context: Context

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "monet_colors")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getColors") {
      if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.S) {
        result.success(null)
      }

      val resources = context.resources
      val colors = mapOf(
        // Accent 1
        "accent1_0" to resources.getColor(android.R.color.system_accent1_0),
        "accent1_10" to resources.getColor(android.R.color.system_accent1_10),
        "accent1_50" to resources.getColor(android.R.color.system_accent1_50),
        "accent1_100" to resources.getColor(android.R.color.system_accent1_100),
        "accent1_200" to resources.getColor(android.R.color.system_accent1_200),
        "accent1_300" to resources.getColor(android.R.color.system_accent1_300),
        "accent1_400" to resources.getColor(android.R.color.system_accent1_400),
        "accent1_500" to resources.getColor(android.R.color.system_accent1_500),
        "accent1_600" to resources.getColor(android.R.color.system_accent1_600),
        "accent1_700" to resources.getColor(android.R.color.system_accent1_700),
        "accent1_800" to resources.getColor(android.R.color.system_accent1_800),
        "accent1_900" to resources.getColor(android.R.color.system_accent1_900),
        "accent1_1000" to resources.getColor(android.R.color.system_accent1_1000),
        // Accent 2
        "accent2_0" to resources.getColor(android.R.color.system_accent2_0),
        "accent2_10" to resources.getColor(android.R.color.system_accent2_10),
        "accent2_50" to resources.getColor(android.R.color.system_accent2_50),
        "accent2_100" to resources.getColor(android.R.color.system_accent2_100),
        "accent2_200" to resources.getColor(android.R.color.system_accent2_200),
        "accent2_300" to resources.getColor(android.R.color.system_accent2_300),
        "accent2_400" to resources.getColor(android.R.color.system_accent2_400),
        "accent2_500" to resources.getColor(android.R.color.system_accent2_500),
        "accent2_600" to resources.getColor(android.R.color.system_accent2_600),
        "accent2_700" to resources.getColor(android.R.color.system_accent2_700),
        "accent2_800" to resources.getColor(android.R.color.system_accent2_800),
        "accent2_900" to resources.getColor(android.R.color.system_accent2_900),
        "accent2_1000" to resources.getColor(android.R.color.system_accent2_1000),
        // Accent 3
        "accent3_0" to resources.getColor(android.R.color.system_accent3_0),
        "accent3_10" to resources.getColor(android.R.color.system_accent3_10),
        "accent3_50" to resources.getColor(android.R.color.system_accent3_50),
        "accent3_100" to resources.getColor(android.R.color.system_accent3_100),
        "accent3_200" to resources.getColor(android.R.color.system_accent3_200),
        "accent3_300" to resources.getColor(android.R.color.system_accent3_300),
        "accent3_400" to resources.getColor(android.R.color.system_accent3_400),
        "accent3_500" to resources.getColor(android.R.color.system_accent3_500),
        "accent3_600" to resources.getColor(android.R.color.system_accent3_600),
        "accent3_700" to resources.getColor(android.R.color.system_accent3_700),
        "accent3_800" to resources.getColor(android.R.color.system_accent3_800),
        "accent3_900" to resources.getColor(android.R.color.system_accent3_900),
        "accent3_1000" to resources.getColor(android.R.color.system_accent3_1000),
        // Neutral 1
        "neutral1_0" to resources.getColor(android.R.color.system_neutral1_0),
        "neutral1_10" to resources.getColor(android.R.color.system_neutral1_10),
        "neutral1_50" to resources.getColor(android.R.color.system_neutral1_50),
        "neutral1_100" to resources.getColor(android.R.color.system_neutral1_100),
        "neutral1_200" to resources.getColor(android.R.color.system_neutral1_200),
        "neutral1_300" to resources.getColor(android.R.color.system_neutral1_300),
        "neutral1_400" to resources.getColor(android.R.color.system_neutral1_400),
        "neutral1_500" to resources.getColor(android.R.color.system_neutral1_500),
        "neutral1_600" to resources.getColor(android.R.color.system_neutral1_600),
        "neutral1_700" to resources.getColor(android.R.color.system_neutral1_700),
        "neutral1_800" to resources.getColor(android.R.color.system_neutral1_800),
        "neutral1_900" to resources.getColor(android.R.color.system_neutral1_900),
        "neutral1_1000" to resources.getColor(android.R.color.system_neutral1_1000),
        // Neutral 2
        "neutral2_0" to resources.getColor(android.R.color.system_neutral2_0),
        "neutral2_10" to resources.getColor(android.R.color.system_neutral2_10),
        "neutral2_50" to resources.getColor(android.R.color.system_neutral2_50),
        "neutral2_100" to resources.getColor(android.R.color.system_neutral2_100),
        "neutral2_200" to resources.getColor(android.R.color.system_neutral2_200),
        "neutral2_300" to resources.getColor(android.R.color.system_neutral2_300),
        "neutral2_400" to resources.getColor(android.R.color.system_neutral2_400),
        "neutral2_500" to resources.getColor(android.R.color.system_neutral2_500),
        "neutral2_600" to resources.getColor(android.R.color.system_neutral2_600),
        "neutral2_700" to resources.getColor(android.R.color.system_neutral2_700),
        "neutral2_800" to resources.getColor(android.R.color.system_neutral2_800),
        "neutral2_900" to resources.getColor(android.R.color.system_neutral2_900),
        "neutral2_1000" to resources.getColor(android.R.color.system_neutral2_1000)
      )
      result.success(colors)
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:monet_colors/monet_colors.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  MonetColors? colors;

  @override
  void initState() {
    super.initState();
    init();
  }

  Future init() async {
    colors = await Monet.get();
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Monet Colors'),
        ),
        body: colors == null
            ? const Center(child: CircularProgressIndicator())
            : ColorsList(colors: colors!),
      ),
    );
  }
}

class ColorsList extends StatelessWidget {
  final MonetColors colors;

  const ColorsList({
    Key? key,
    required this.colors,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const SizedBox(height: 24),
        ColorItem(color: colors.primary[0]!, label: 'Primary [0]'),
        ColorItem(color: colors.primary[10]!, label: 'Primary [10]'),
        ColorItem(color: colors.primary[50]!, label: 'Primary [50]'),
        ColorItem(color: colors.primary[100]!, label: 'Primary [100]'),
        ColorItem(color: colors.primary[200]!, label: 'Primary [200]'),
        ColorItem(color: colors.primary[300]!, label: 'Primary [300]'),
        ColorItem(color: colors.primary[400]!, label: 'Primary [400]'),
        ColorItem(color: colors.primary[500]!, label: 'Primary [500]'),
        ColorItem(color: colors.primary[600]!, label: 'Primary [600]'),
        ColorItem(color: colors.primary[700]!, label: 'Primary [700]'),
        ColorItem(color: colors.primary[800]!, label: 'Primary [800]'),
        ColorItem(color: colors.primary[900]!, label: 'Primary [900]'),
        ColorItem(color: colors.primary[1000]!, label: 'Primary [1000]'),
        const SizedBox(height: 24),
        ColorItem(color: colors.secondary[0]!, label: 'Secondary [0]'),
        ColorItem(color: colors.secondary[10]!, label: 'Secondary [10]'),
        ColorItem(color: colors.secondary[50]!, label: 'Secondary [50]'),
        ColorItem(color: colors.secondary[100]!, label: 'Secondary [100]'),
        ColorItem(color: colors.secondary[200]!, label: 'Secondary [200]'),
        ColorItem(color: colors.secondary[300]!, label: 'Secondary [300]'),
        ColorItem(color: colors.secondary[400]!, label: 'Secondary [400]'),
        ColorItem(color: colors.secondary[500]!, label: 'Secondary [500]'),
        ColorItem(color: colors.secondary[600]!, label: 'Secondary [600]'),
        ColorItem(color: colors.secondary[700]!, label: 'Secondary [700]'),
        ColorItem(color: colors.secondary[800]!, label: 'Secondary [800]'),
        ColorItem(color: colors.secondary[900]!, label: 'Secondary [900]'),
        ColorItem(color: colors.secondary[1000]!, label: 'Secondary [1000]'),
        const SizedBox(height: 24),
        ColorItem(color: colors.tertiary[0]!, label: 'Tertiary [0]'),
        ColorItem(color: colors.tertiary[10]!, label: 'Tertiary [10]'),
        ColorItem(color: colors.tertiary[50]!, label: 'Tertiary [50]'),
        ColorItem(color: colors.tertiary[100]!, label: 'Tertiary [100]'),
        ColorItem(color: colors.tertiary[200]!, label: 'Tertiary [200]'),
        ColorItem(color: colors.tertiary[300]!, label: 'Tertiary [300]'),
        ColorItem(color: colors.tertiary[400]!, label: 'Tertiary [400]'),
        ColorItem(color: colors.tertiary[500]!, label: 'Tertiary [500]'),
        ColorItem(color: colors.tertiary[600]!, label: 'Tertiary [600]'),
        ColorItem(color: colors.tertiary[700]!, label: 'Tertiary [700]'),
        ColorItem(color: colors.tertiary[800]!, label: 'Tertiary [800]'),
        ColorItem(color: colors.tertiary[900]!, label: 'Tertiary [900]'),
        ColorItem(color: colors.tertiary[1000]!, label: 'Tertiary [1000]'),
        const SizedBox(height: 24),
        ColorItem(color: colors.neutral[0]!, label: 'Neutral [0]'),
        ColorItem(color: colors.neutral[10]!, label: 'Neutral [10]'),
        ColorItem(color: colors.neutral[50]!, label: 'Neutral [50]'),
        ColorItem(color: colors.neutral[100]!, label: 'Neutral [100]'),
        ColorItem(color: colors.neutral[200]!, label: 'Neutral [200]'),
        ColorItem(color: colors.neutral[300]!, label: 'Neutral [300]'),
        ColorItem(color: colors.neutral[400]!, label: 'Neutral [400]'),
        ColorItem(color: colors.neutral[500]!, label: 'Neutral [500]'),
        ColorItem(color: colors.neutral[600]!, label: 'Neutral [600]'),
        ColorItem(color: colors.neutral[700]!, label: 'Neutral [700]'),
        ColorItem(color: colors.neutral[800]!, label: 'Neutral [800]'),
        ColorItem(color: colors.neutral[900]!, label: 'Neutral [900]'),
        ColorItem(color: colors.neutral[1000]!, label: 'Neutral[1000]'),
        const SizedBox(height: 24),
        ColorItem(
          color: colors.neutralVariant[0]!,
          label: 'Neutral Variant [0]',
        ),
        ColorItem(
          color: colors.neutralVariant[10]!,
          label: 'Neutral Variant [10]',
        ),
        ColorItem(
          color: colors.neutralVariant[50]!,
          label: 'Neutral Variant [50]',
        ),
        ColorItem(
          color: colors.neutralVariant[100]!,
          label: 'Neutral Variant [100]',
        ),
        ColorItem(
          color: colors.neutralVariant[200]!,
          label: 'Neutral Variant [200]',
        ),
        ColorItem(
          color: colors.neutralVariant[300]!,
          label: 'Neutral Variant [300]',
        ),
        ColorItem(
          color: colors.neutralVariant[400]!,
          label: 'Neutral Variant [400]',
        ),
        ColorItem(
          color: colors.neutralVariant[500]!,
          label: 'Neutral Variant [500]',
        ),
        ColorItem(
          color: colors.neutralVariant[600]!,
          label: 'Neutral Variant [600]',
        ),
        ColorItem(
          color: colors.neutralVariant[700]!,
          label: 'Neutral Variant [700]',
        ),
        ColorItem(
          color: colors.neutralVariant[800]!,
          label: 'Neutral Variant [800]',
        ),
        ColorItem(
          color: colors.neutralVariant[900]!,
          label: 'Neutral Variant [900]',
        ),
        ColorItem(
          color: colors.neutralVariant[1000]!,
          label: 'Neutral Variant [1000]',
        ),
        const SizedBox(height: 24),
      ],
    );
  }
}

class ColorItem extends StatelessWidget {
  final Color color;
  final String label;

  const ColorItem({
    Key? key,
    required this.color,
    required this.label,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 24,
      width: double.infinity,
      color: color,
      child: Align(
        alignment: AlignmentDirectional.centerStart,
        child: Padding(
          padding: const EdgeInsetsDirectional.only(start: 8),
          child: Text(label),
        ),
      ),
    );
  }
}
